libqtile/backend/base/window.py

```diff
@@ -28,6 +29,8 @@ class FloatStates(enum.Enum):
 
 
 class _Window(CommandObject, metaclass=ABCMeta):
+    _focus_filters: list[Callable] = []
```

Comment:
I still don't really love having some class-level thing and super indirection that is only called from one place.

The prompt uses widget_grab_keyboard() in the bar, which sets `_has_keyboard` among other things. Can we iterate over all bars instead of all widets? That seems cheaper, and won't be prompt-specific (although prompt is the only caller of widget_grab_keyboard() in the tree, currently).

Reply:
Alternatively, `qtile.current_window` should be the result of `create_internal()` for Bar windows, so you can probably just get away with checking if that is a `window.Internal` window, which won't need any iteration at all.

Reply:
...and that will help with focus being stolen from all the places that use `create_internal()`, which seems good? Right now it's:
```
libqtile/bar.py:                self.window = qtile.core.create_internal(  # type: ignore [call-arg]
libqtile/bar.py:                self.window = qtile.core.create_internal(self.x, self.y, width, height)
libqtile/layout/tree.py:        self._panel = self.group.qtile.core.create_internal(
libqtile/popup.py:        self.win: Any = qtile.core.create_internal(
```
So probably we don't want to steal focus from our own internal popups?

Reply:
Imho it's mostly about where you want to have the code:
Your approach keeps it central (which may lead to more imports in the window module sooner or later), whereas mine attempts to put it where the widget is defined.
E.g. the internal route may become problematic once the first internal window decides it wants to pass focus and then you'll have to import the prompt widget in the Window class...

Also, I'm not 100% sure #537 could be considered "fixed" with only the unofficial way of re-writing `can_steal_focus()`, which I consider a rather ugly hack.

Reply:
I think we're misunderstanding each other, so let me please clarify:

> Also, I'm not 100% sure https://github.com/qtile/qtile/issues/537 could be considered "fixed" with only the unofficial way of re-writing can_steal_focus(), which I consider a rather ugly hack.

I was talking about the current way of doing
```python
@hook.subscribe.client_new
def prevent_focus_steal(client):
    client.__class__.can_steal_focus = property(myfunc)
```
in the `config.py` is ugly and therefore imho doesn't fix #537.

The proposed way of doing
```python
from libqtile.backend.base import Window

Window.register_focus_filter(myfunc)
```
is clean and therefore imo fixes #537.

Also, if you don't want windows from `create_internal()` to be stolen focus from, you can do something like
```python
def create_internal(..., prevent_focus_steal=False):
...
    if prevent_focus_steal:
        Window.register_focus_filter(lambda win: self.qtile.current_window == win)
...
```
or just do it at a later point in the bar class.

I think it's important to understand that `can_steal_focus()` defines whether a newly created window can steal focus from others, but in our case it's _another_ already created window which wants to prevent _all_ newly created windows from stealing focus from it. This will always require some sort of registration with the `can_steal_focus()` function executed by other windows.

Reply:
Oh, maybe I missed your point here:

> I think it's important to understand that can_steal_focus() defines whether a newly created window can steal focus from others, but in our case it's another already created window which wants to prevent all newly created windows from stealing focus from it. This will always require some sort of registration with the can_steal_focus() function executed by other windows.

What we really want then, is some kind of can_be_stolen_from(), which is what #537 wants too. Can we use the same trick, just looking at `qtile.current_window.can_be_stolen_from()`, which defaults to true, but can be overridden in specific cases, or by users as they see fit?